tests/python/pants_test/engine/test_isolated_process.py

```diff
@@ -89,21 +98,26 @@ def create_cat_stdout_rules():
   ]
 
 
-class JavacVersionExecutionRequest(datatype('JavacVersionExecutionRequest', ['bin_path'])):
+@typed_data(BinaryLocation)
+class JavacVersionExecutionRequest:
+
+  @property
+  def bin_path(self):
+    return self.binary_location.primitive__str
```

Comment:
This is not great syntax, unfortunately. I think this might have gone too far when it removed field names.

I would propose that the right way to land this would be to simplify it back to:
```
class JavacVersionExecutionRequest(datatype('JavacVersionExecutionRequest', [('bin_path', str)])):
```

IE: It should be straightforward to use the `datatype` name and accept tuples in the field list. It should also be possible to remove the classname parameter from `datatype` and just make an anonymous class, on the assumption that it will be extended. Then you'd have:
```
class JavacVersionExecutionRequest(datatype([('bin_path', str)])):
```

Reply:
Ok, this is pretty clear reasoning. I think you made this syntax clear in your comments above and I just wasn't sure what the end result would look like. The only utility of the decorator was to avoid having to spell out the same class name again, which is clear to me how to address from your description above. I really like the result you've proposed (especially the pretty natural integration with existing `datatype()` uses, so thanks for taking the time to make this clear.

Reply:
I implemented your proposal (which was great, btw) in 9e8055f (and was able to mostly keep the testing I already wrote intact), and changed every `datatype()` call site to not specify a class name in 2d1489e. Let me know what you think.

Reply:
After rebasing, the shas are bd3a2d6 and 6420f5b.